fs-storage/src/file_storage.rs

```diff
@@ -8,51 +9,110 @@ use std::{
     path::{Path, PathBuf},
 };
 
+use crate::base_storage::BaseStorage;
 use data_error::{ArklibError, Result};
 
 const STORAGE_VERSION: i32 = 2;
 const STORAGE_VERSION_PREFIX: &str = "version ";
 
-pub struct FileStorage {
+pub struct FileStorage<K, V> {
     label: String,
     path: PathBuf,
     timestamp: SystemTime,
+    pub value_by_id: BTreeMap<K, V>,
 }
 
-impl FileStorage {
+impl<K, V> FileStorage<K, V>
+where
+    K: serde::Serialize,
+    V: serde::Serialize,
+{
     /// Create a new file storage with a diagnostic label and file path
     pub fn new(label: String, path: &Path) -> Self {
         Self {
             label,
             path: PathBuf::from(path),
             timestamp: SystemTime::now(),
+            value_by_id: BTreeMap::new(),
         }
     }
 
-    /// Check if underlying file has been updated
-    ///
-    /// This check can be used before reading the file.
-    pub fn is_file_updated(&self) -> Result<bool> {
+    /// Verify the version stored in the file header
+    fn verify_version(&self, header: &str) -> Result<()> {
+        if !header.starts_with(STORAGE_VERSION_PREFIX) {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Unknown storage version prefix".to_owned(),
+            ));
+        }
+
+        let version = header[STORAGE_VERSION_PREFIX.len()..]
+            .parse::<i32>()
+            .map_err(|_err| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to parse storage version".to_owned(),
+                )
+            })?;
+
+        if version != STORAGE_VERSION {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                format!(
+                    "Storage version mismatch: expected {}, found {}",
+                    STORAGE_VERSION, version
+                ),
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+impl<K, V> BaseStorage<K, V> for FileStorage<K, V>
+where
+    K: FromStr
+        + Hash
+        + Eq
+        + Ord
+        + Debug
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned,
+    V: Debug + Clone + serde::Serialize + serde::de::DeserializeOwned,
+{
+    fn get(&self, id: &K) -> Option<V> {
+        self.value_by_id.get(id).cloned()
+    }
+
+    fn set(&mut self, id: K, value: V) -> Result<()> {
+        self.value_by_id.insert(id, value);
+        self.timestamp = std::time::SystemTime::now();
+        Ok(())
+    }
```

Comment:
Do you think we should also write to disk here?
        Since we are updating the map, we should also update the information on file system

Reply:
Good question. We probably should write to disk periodically, not immediately. This is discussable.

Basically, updates could come often enough to produce several writes simultaneously. While we should be protected from dirty writes thanks to [fs-atomic-versions](https://github.com/ARK-Builders/ark-rust/tree/main/fs-atomic-versions), we still want to avoid unnecessary parallelism and batch updates together and write them once in a while, each 1 second.

On Kotlin side, this was achieved using [debounce](https://github.com/ARK-Builders/arklib-android/blob/377734870453889f3e9a8821bb1cac3f7f299ab8/lib/src/main/java/dev/arkbuilders/arklib/data/storage/BaseStorage.kt#L31)

Reply:
cc @twitu

Reply:
The proper way to do this would be to create an asynchronous job that periodically writes to disk. This is exactly what is done using debounce in Kotlin. However, adding async support will be outside the scope of this PR. Thankfully [async traits](https://blog.rust-lang.org/inside-rust/2023/05/03/stabilizing-async-fn-in-trait.html) have recently been stabilized in Rust. So it can be added in separately.

For now the safe but less efficient method will be to flush to disk on each write. On the other hand we can go for a slightly less safe but more efficient method by using a BufferedWriter to write data to the file. The BufferedWriter will automatically batch writes and flush to disk.

I propose going ahead with BufferedWriter approach.